pkg/filetree/filetree.go

```diff
@@ -0,0 +1,749 @@
+			if err != nil {
+				return err
+			}
+			if fn != nil {
+				break
+			}
+			pathsToAdd = append(pathsToAdd, parentPaths[idx])
+		}
+
+		// add each path with no file reference; add these in sorted path order (which is guaranteed to be
+		// the reverse of the order of insertion)
+		for idx := len(pathsToAdd) - 1; idx >= 0; idx-- {
+			newFn := filenode.NewDir(pathsToAdd[idx], nil)
+			if err = t.setFileNode(newFn); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// setFileNode adds the given path to the Tree with the specific file.Reference.
+func (t *FileTree) setFileNode(fn *filenode.FileNode) error {
+	if fn == nil {
+		return fmt.Errorf("must provide a FileNode when adding paths")
+	}
+
+	if existingNode := t.tree.Node(filenode.IDByPath(fn.RealPath)); existingNode != nil {
+		return t.tree.Replace(existingNode, fn)
+	}
+
+	parentPath, err := fn.RealPath.ParentPath()
+	if err != nil {
+		return fmt.Errorf("unable to determine parent path while adding path=%q: %w", fn.RealPath, err)
+	}
+
+	parentNode, err := t.node(parentPath, linkResolutionStrategy{})
+	if err != nil {
+		return err
+	}
+	if parentNode == nil {
+		return fmt.Errorf("unable to find parent path=%q while adding path=%q", parentPath, fn.RealPath)
+	}
+
+	return t.tree.AddChild(parentNode, fn)
+}
+
+// RemovePath deletes the file.Reference from the FileTree by the given path. If the basename of the given path
+// is a symlink then the symlink is removed (not the destination of the symlink). If the path does not exist, this is a
+// nop.
+func (t *FileTree) RemovePath(path file.Path) error {
+	if path.Normalize() == "/" {
+		return ErrRemovingRoot
+	}
+
+	fn, err := t.node(path, linkResolutionStrategy{
+		FollowAncestorLinks: true,
+		FollowBasenameLinks: false,
+	})
+	if err != nil {
+		return err
+	}
+	if fn == nil {
+		return nil
+	}
+
+	_, err = t.tree.RemoveNode(fn)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// RemoveChildPaths deletes all children of the given path (not including the given path). Note: if the given path
+// basename is a symlink, then the symlink is followed before resolving children. If the path does not exist, this is a
+// nop.
+func (t *FileTree) RemoveChildPaths(path file.Path) error {
+	fn, err := t.node(path, linkResolutionStrategy{
+		FollowAncestorLinks: true,
+		FollowBasenameLinks: true,
+	})
+	if err != nil {
+		return err
+	}
+	if fn == nil {
+		// can't remove child paths for Node that doesn't exist!
+		return nil
+	}
+	for _, child := range t.tree.Children(fn) {
+		_, err := t.tree.RemoveNode(child)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Reader returns a tree.Reader useful for Tree traversal.
+func (t *FileTree) Reader() tree.Reader {
+	return t.tree
+}
+
+// PathDiff shows the path differences between two trees (useful for testing)
+func (t *FileTree) PathDiff(other *FileTree) (extra, missing []file.Path) {
+	extra = make([]file.Path, 0)
+	missing = make([]file.Path, 0)
```

Comment:
These two lines ( for `extra` and `missing`) might be unnecessary, since no capacity is allocated, and these variables are already declared via the return parameters.